### PR TITLE
[#9382] Migrate response comments back-end

### PR DESCRIPTION
**Pull request by wkurniawan07**
Part of #9382 

This migrates all the back-end functions related to comments (except for `GET`, which is still coupled to some other actions) to REST APIs, endpoint: `/responsecomment`.

The output format is still tentative as it is still not known yet what the front-end requires. The request format should be final, but no guarantee.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import teammates.common.exception.EntityNotFoundException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.Assumption;
+import teammates.common.util.Const;
+import teammates.ui.webapi.output.FeedbackResponseCommentData;
+
+/**
+ * Edits a feedback response comment.
+ */
+public class EditFeedbackResponseCommentAction extends Action {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java]: `SaveFeedbackResponseCommentAction`?

File: src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    protected AuthType getMinAuthLevel() {
+        return AuthType.LOGGED_IN;
+    }
+
+    @Override
+    public void checkSpecificAccessControl() {
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String feedbackResponseId = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_RESPONSE_ID);
+        Long feedbackResponseCommentId = getLongRequestParamValue(Const.ParamsNames.FEEDBACK_RESPONSE_COMMENT_ID);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java]: The `courseId`, `feedbackSessionName` and `feedbackResponseId` are not necessary as they all stored in `FeedbackResponseAttributes`? Same for Create/Delete action.

File: src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+        FeedbackSessionAttributes session = logic.getFeedbackSession(feedbackSessionName, courseId);
+        FeedbackResponseAttributes response = logic.getFeedbackResponse(feedbackResponseId);
+        Assumption.assertNotNull(response);
+
+        // Edit comment text
+        String commentText = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_RESPONSE_COMMENT_TEXT);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java]: Are we supposed to use request body here? Same for create request?

Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/action/EditFeedbackResponseCommentAction.java]: You're looking at an outdated diff...

File: src/main/java/teammates/ui/webapi/action/ActionFactory.java [hunk 4]
```diff
@@ -48,6 +48,9 @@
         map(ResourceURIs.INSTRUCTORS, DELETE, DeleteInstructorAction.class);
         map(ResourceURIs.INSTRUCTOR, GET, GetInstructorAction.class);
         map(ResourceURIs.INSTRUCTOR_PRIVILEGE, GET, GetInstructorPrivilegeAction.class);
+        map(ResourceURIs.RESPONSE_COMMENT, POST, CreateFeedbackResponseCommentAction.class);
+        map(ResourceURIs.RESPONSE_COMMENT, PUT, SaveFeedbackResponseCommentAction.class);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/ActionFactory.java]: Should be `UpdateFeedbackResponseCommentAction` now LOL. Take your time to change hahahah.

File: src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    @Override
+    public ActionResult execute() {
+        String feedbackResponseId = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_RESPONSE_ID);
+
+        FeedbackResponseAttributes response = logic.getFeedbackResponse(feedbackResponseId);
+        Assumption.assertNotNull(response);
+
+        FeedbackResponseCommentSaveRequest comment = getAndValidateRequestBody(FeedbackResponseCommentSaveRequest.class);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java]: Shouldn't this be `FeedbackResponseCommentCreateRequest`?

File: src/main/java/teammates/ui/webapi/action/DeleteFeedbackResponseCommentAction.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+
+        if (instructor != null && frc.commentGiver.equals(instructor.email)) { // giver, allowed by default
+            return;
+        }
+
+        // Case 1: the delete request comes from instructor
+
+        if (instructor != null) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/DeleteFeedbackResponseCommentAction.java]: We might need something like `intent`  or just use `frc.isCommentFromFeedbackParticipant` to distinguish request from instructor or student? A person can be both instructor and student (this is true for demo data). In this case, he will never be able to delete his comment for responses he has given.

File: src/main/java/teammates/ui/webapi/request/FeedbackResponseCommentSaveRequest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * The save request of a feedback response comment.
+ */
+public class FeedbackResponseCommentSaveRequest extends BasicRequest {
+
+    private String commentText;
+    @Nullable
+    private String showCommentTo;
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/request/FeedbackResponseCommentSaveRequest.java]: I don't like the current way of using string for `showCommentTo` and split it to construct feedbackPariticpantType. At least it should be something like `private List<FeedbackParticipantType> showCommentTo`. I think currently, we will leave it there and fix it when we migrate the comment feature in frontend.

File: src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java (hunk 5 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java]: It's used for both creation and update.

File: src/main/java/teammates/ui/webapi/action/DeleteFeedbackResponseCommentAction.java (hunk 6 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/action/DeleteFeedbackResponseCommentAction.java]: > he will never be able to delete his comment for responses he has given

How is this not covered in the access control check before this? (giver, allowed by default)

File: src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java (hunk 5 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/CreateFeedbackResponseCommentAction.java]: Shouldn't it be more clear to have `createRequest` and `updateRequest` separately?

**Pull request merged**